src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs

```diff
@@ -163,5 +171,63 @@ public static void AppendStatusTrailers(HttpResponse response, Status status)
             response.AppendTrailer(GrpcProtocolConstants.StatusTrailer, status.StatusCode.ToTrailerString());
             response.AppendTrailer(GrpcProtocolConstants.MessageTrailer, status.Detail);
         }
+
+        public static AuthContext CreateAuthContext(X509Certificate2 clientCertificate)
+        {
+            // Map X509Certificate2 values to AuthContext. The name/values come BoringSSL via C Core
+            // https://github.com/grpc/grpc/blob/a3cc5361e6f6eb679ccf5c36ecc6d0ca41b64f4f/src/core/lib/security/security_connector/ssl_utils.cc#L206-L248
+
+            var properties = new Dictionary<string, List<AuthProperty>>();
+
+            string peerIdentityPropertyName = null;
+
+            foreach (var extension in clientCertificate.Extensions)
+            {
+                if (extension.Oid.Value == GrpcProtocolConstants.X509SubjectAlternativeNameId)
+                {
+                    var formattedAlternativeNamesText = extension.Format(multiLine: true);
+
+                    // The format of subject names changes between operating systems
+                    // Parse subject names using a regex that supports all major operating systems (Windows/Mac/Linux)
+                    foreach (Match match in SubjectAlternativeNameRegex.Matches(formattedAlternativeNamesText))
```

Comment:
Seriously, no regular expressions. SAN strings are separated by semicolons, and prefixed by DNS=. Split, and substring.

Reply:
Also SANs are for HTTPS, not for client certs. This feels a bit like a misuse for client certificate authentication, if this is what's happening here.

Reply:
It is different between OSes.

DNS Name=xxx separated by new lines on Windows
DNS:xxx separated by commas on Mac/Linux

I could parse it without a regex if you want, but I don't think it impacts security. Removing the regex would mostly be for performance.

Reply:
Keep it, but it's horrible. @bartonjs should be bugged to try to make this all cross plat identical.

Reply:
https://github.com/dotnet/corefx/issues/22068

Reply:
This code will fail on a ja-JP machine, btw.  (Windows localizes "DNS Name")

Reply:
https://github.com/dotnet/wcf/blob/08371d989fc1d230dbe9520916644f7a7f5dc954/src/System.Private.ServiceModel/src/System/IdentityModel/Claims/X509CertificateClaimSet.cs#L297-L332  and https://github.com/dotnet/wcf/blob/08371d989fc1d230dbe9520916644f7a7f5dc954/src/System.Private.ServiceModel/src/System/IdentityModel/Claims/X509CertificateClaimSet.cs#L376-L484 are a working example of how to do this.

Reply:
Blarg.

Thanks for the links. At least @blowdart will get to see the regex disappear 😋